aries_cloudagent/connections/models/diddoc_v2/diddoc.py

```diff
@@ -0,0 +1,311 @@
+
+    def __init__(
+        self,
+        id: str,
+        alsoKnownAs: list = None,
+        controller=None,
+        verificationMethod: list = [],
+        authentication: list = [],
+        assertionMethod: list = [],
+        keyAgreement: list = [],
+        capabilityInvocation: list = [],
+        capabilityDelegation: list = [],
+        publicKey: list = [],
+        service: list = [],
+    ) -> None:
+
+        """
+        Initialize the DIDDoc instance.
+
+        Retain DID ('id' in DIDDoc context); initialize verification keys
+        and services to empty lists.
+
+        Args:
+            did: DID for current DIDDoc.
+            id: DIDDoc id.
+            alsoKnownAs: One or more other identifiers of the DIDDoc.
+            controller: Contain verification relationships of the DIDDoc.
+            verificationMethod: Specific verification method of the DIDDoc.
+            authentication: Specific verification method of the DIDDoc.
+            assertionMethod: Specific verification method of the DIDDoc.
+            keyAgreement: Specific verification method of the DIDDoc.
+            capabilityInvocation: Specific verification method of the DIDDoc.
+            capabilityDelegation: Specific verification method of the DIDDoc.,
+            publicKey: Specific verification method of the DIDDoc.
+            service: Communicating of the DID subject or associated entities.
+
+        Raises:
+            ValueError: for bad input DID.
+
+        """
+
+        self._id = id
+        self._alsoKnownAs = alsoKnownAs
+        self._controller = controller
+        self._index = {}
+        self._ref_content = {}
+
+        params = (
+            ("verificationMethod", verificationMethod),
+            ("authentication", authentication),
+            ("assertionMethod", assertionMethod),
+            ("keyAgreement", keyAgreement),
+            ("capabilityInvocation", capabilityInvocation),
+            ("capabilityDelegation", capabilityDelegation),
+            ("publicKey", publicKey),
+            ("service", service),
+        )
+
+        for param in params:
+            aux_content = []
+            for item in param[1]:
+                if not isinstance(item, str):
+                    did_item = self._index.get(item)
+                    if not self._index.get(item):
+                        self._index[item.id] = item  # {id: <kind of param>}
+                        aux_content.append(item.id)
+                    else:
+                        if not (did_item.serialize() == item.serialize()):
+                            raise ValueError(
+                                "{} has different specifications".format(item.id)
+                            )
+                else:
+                    if not self._index.get(item):
+                        self._index[item] = param[0]
+                        aux_content.append(item.id)
+            self._ref_content[param[0]] = aux_content
```

Comment:
Some comments describing this block of code would be greatly appreciated. Currently, the flow is difficult to follow and it's function is not immediately clear. I would consider making this a separate helper function.